File: app/src/main/java/fr/gstraymond/db/json/JsonList.java
package fr.gstraymond.db.json;

import android.content.Context;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.magic.card.search.commons.json.MapperUtil;
import com.magic.card.search.commons.log.Log;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.util.ArrayList;
import java.util.List;

import fr.gstraymond.search.model.response.Card;

public class JsonList {

    private Log log = new Log(this);

    private Context context;
    private MapperUtil<List<Card>> mapperUtil;
    private String listName;
    private List<Card> cards = new ArrayList<>();

    public JsonList(Context context, ObjectMapper objectMapper, String listName) {
        this.context = context;
        this.mapperUtil = MapperUtil.fromCollectionType(objectMapper, Card.class);
        this.listName = "lists_" + listName;
        this.cards = load();
    }

    private List<Card> load() {
        try {
            FileInputStream inputStream = context.openFileInput(listName);
            return new ArrayList<>(mapperUtil.read(inputStream));
        } catch (FileNotFoundException e) {
            log.w("get: %s", e);
            save();
            return cards;
        }
    }

    public void save() {
        try {
            FileOutputStream fos = context.openFileOutput(listName, Context.MODE_PRIVATE);
            fos.write(mapperUtil.asJsonString(cards).getBytes());
            fos.close();
        } catch (Exception e) {
            log.e("save", e);
        }
    }

    public boolean addOrRemove(Card card) {
        boolean contains = contains(card);
        if (!contains) cards.add(card);
        else cards.remove(card);
        save();
        log.d("addOrRemove %s -> removed? %s", card, contains);
        return !contains;
    }

    public boolean contains(Card card) {
        String id = getId(card);
        for (Card currentCard : cards) {
            if (getId(currentCard).equals(id)) return true;
        }
        return false;
    }

    private String getId(Card card) {
        return String.format("%s %s %s", card.getTitle(), card.getType(), card.getCastingCost());
    }

    public List<Card> getCards() {
        return cards;
    }
}

File: app/src/main/kotlin/fr/gstraymond/db/json/JsonList.kt
package fr.gstraymond.db.json

import android.content.Context
import android.content.Context.MODE_APPEND
import android.content.Context.MODE_PRIVATE
import com.magic.card.search.commons.json.MapperUtil
import java.io.FileNotFoundException
import java.io.FileOutputStream
import kotlin.concurrent.thread

abstract class JsonList<A>(private val context: Context,
                           private val mapperUtil: MapperUtil<A>,
                           listPrefix: String = "list",
                           listName: String,
                           loadOnInit: Boolean = true) : MemoryUidList<A>() {

    protected val listName = "${listPrefix}_$listName"

    override val elems: MutableList<A> = mutableListOf()
    override val index: MutableMap<String, A> = mutableMapOf()

    init {
        if (loadOnInit) init()
    }

    protected fun init() {
        elems.addAll(
                try {
                    mutableListOf<A>().apply {
                        context.openFileInput(listName).bufferedReader().useLines {
                            it.forEach {
                                mapperUtil.read(it)?.let { add(it) }
                            }
                        }
                    }
                } catch (e: FileNotFoundException) {
                    log.w("get: %s", e)
                    //save(listOf())
                    mutableListOf<A>()
                })

        index.putAll(
                mutableMapOf(*(elems.map { it.uid() to it }.toTypedArray()))
        )
    }

    override fun save(elements: List<A>) {
        super.save(elements)
        writeAll()
    }

    override fun delete(elem: A) = super.delete(elem).apply { writeAll() }

    override fun update(elem: A) = super.update(elem).apply { writeAll() }

    override fun clear() = super.clear().apply {
        try {
            context.deleteFile(listName)
        } catch (e: Exception) {
            log.e("clear: $listName", e)
        }
    }

    override fun append(elem: A) = super.append(elem).apply {
        synchronized(this) {
            write(MODE_APPEND) { it.write(elem) }
        }
    }

    private fun write(mode: Int, f: (FileOutputStream) -> Unit) {
        try {
            context.openFileOutput(listName, mode).use { f(it) }
        } catch (e: Exception) {
            log.e("save", e)
        }
    }

    private fun writeAll() {
        thread {
            synchronized(this) {
                log.d("write $javaClass")
                write(MODE_PRIVATE) { stream ->
                    elems.map { stream.write(it) }
                }
            }
        }
    }

    private fun FileOutputStream.write(elem: A) {
        write("${mapperUtil.asJsonString(elem)}\n".toByteArray())
    }
}
